java/source/io.rst

```diff
@@ -0,0 +1,646 @@
+                VectorSchemaRoot vectorSchemaRootRecover = reader.getVectorSchemaRoot();
+                System.out.print(vectorSchemaRootRecover.contentToTSVString());
+            }
+        } catch (IOException e) {
+            e.printStackTrace();
+        }
+    }
+
+.. testoutput::
+
+    Record batches readed: 3
+    name    age
+    David    10
+    Gladis    20
+    Juan    30
+    name    age
+    Nidia    15
+    Alexa    20
+    Mara    15
+    name    age
+    Raul    34
+    Jhon    29
+    Thomy    33
+
+Reading Streaming Format
+************************
+
+Read - From File
+----------------
+
+.. testcode::
+
+    import org.apache.arrow.memory.RootAllocator;
+    import org.apache.arrow.vector.ipc.ArrowStreamReader;
+    import org.apache.arrow.vector.VectorSchemaRoot;
+    import java.io.File;
+    import java.io.FileInputStream;
+    import java.io.IOException;
+
+    try(RootAllocator rootAllocator = new RootAllocator(Long.MAX_VALUE)) {
+        File file = new File("./thirdpartydeps/arrowfiles/data2.arrow");
+        try (FileInputStream fileInputStreamForStream = new FileInputStream(file);
+             ArrowStreamReader reader = new ArrowStreamReader(fileInputStreamForStream, rootAllocator)) {
+            while (reader.loadNextBatch()) {
+                VectorSchemaRoot vectorSchemaRootRecover = reader.getVectorSchemaRoot();
+                System.out.print(vectorSchemaRootRecover.contentToTSVString());
+            }
+        } catch (IOException e) {
+            e.printStackTrace();
+        }
+    }
+
+.. testoutput::
+
+    name    age
+    David    10
+    Gladis    20
+    Juan    30
+    name    age
+    Nidia    15
+    Alexa    20
+    Mara    15
+    name    age
+    Raul    34
+    Jhon    29
+    Thomy    33
+
+Read - From Buffer
+------------------
+
+.. testcode::
+
+    import org.apache.arrow.memory.RootAllocator;
+    import org.apache.arrow.vector.ipc.ArrowStreamReader;
+
+    import java.io.ByteArrayInputStream;
+    import java.io.IOException;
+    import java.nio.file.Files;
+    import java.nio.file.Path;
+    import java.nio.file.Paths;
+
+    try(RootAllocator rootAllocator = new RootAllocator(Long.MAX_VALUE)) {
+        Path path = Paths.get("./thirdpartydeps/arrowfiles/data2.arrow");
+        try (ArrowStreamReader reader = new ArrowStreamReader(new ByteArrayInputStream(Files.readAllBytes(path)), rootAllocator)){
+            while(reader.loadNextBatch()){
+                System.out.print(reader.getVectorSchemaRoot().contentToTSVString());
+            }
+        } catch (IOException e) {
+            e.printStackTrace();
+        }
+    }
+
+.. testoutput::
+
+    name    age
+    David    10
+    Gladis    20
+    Juan    30
+    name    age
+    Nidia    15
+    Alexa    20
+    Mara    15
+    name    age
+    Raul    34
+    Jhon    29
+    Thomy    33
+
+Reading Parquet File
+********************
+
+Please check :ref:`arrow-dataset`
+
+Writing Fresh Data
```

Comment:
These two examples aren't really part of IO per se. This is about the basic usage of the Arrow Java library: roots and vectors are mutable.

Reply:
I'd rather have examples in the basic section that show how to use VectorLoader/VectorUnloader, and how to modify the data in the root (I suppose we have that one already). If we have to show a combined example, I'd rather modify the existing write examples (e.g. they could fill a vector with random integers on each iteration).

Reply:
We are going to add examples for VectorLoader/VectorUnloader, and how to modify the data in the root.

What about if maintain this as it is, because these could be help something when need to persist fresh data thru io files.

Reply:
To me, the recipe feels very unnatural. For instance, why not just have this in the main recipe up above? Then we don't need two separate recipes.

```
                // Delete the set() calls above
                writer.start();
                for (int i = 0; i < 3; i++) {
                    nameVector.set(0, ("New-"+(i+1)).getBytes());
                    nameVector.set(1, ("New-"+(i+1)).getBytes());
                    nameVector.set(2, ("New-"+(i+1)).getBytes());
                    ageVector.set(0, (i+2)*10);
                    ageVector.set(1, (i+2)*20);
                    ageVector.set(2, (i+2)*30);
                    writer.writeBatch();
                }
```

Reply:
The last example we could keep, but we should create the writer after the reader and use the reader's schema to create the writer's root.

Reply:
Yes, but in the last case the data flow is:

- First the user have partial data to write
- Second the user obtain more data from a external file a add this data to the initial data

Please let me know if this is ok

Reply:
To me that is a very specific use case, and it should be clear from general principles. We can't cover every possible combination of things people want to do. So I think we should to break down our examples into very focused parts, so that they can be applied in more situations, rather than trying to make examples for every situation we can think of.

Reply:
Yes is difficult to cover all scenarios in our cookbook. We could improve and cover more custom escenarios in next versions.

Please let me know if we could continue as it is.

Reply:
I think if we removed other examples of updating and writing data we should remove this one too. We can focus on just how to update the root and use vectorloader/vectorunloader in separate recipes.